compiler/rustc_mir_transform/src/remove_uninit_drops.rs

```diff
@@ -0,0 +1,171 @@
+use rustc_index::bit_set::BitSet;
+use rustc_middle::mir::{Body, Field, Rvalue, Statement, StatementKind, TerminatorKind};
+use rustc_middle::ty::subst::SubstsRef;
+use rustc_middle::ty::{self, ParamEnv, Ty, TyCtxt, VariantDef};
+use rustc_mir_dataflow::impls::MaybeInitializedPlaces;
+use rustc_mir_dataflow::move_paths::{LookupResult, MoveData, MovePathIndex};
+use rustc_mir_dataflow::{self, move_path_children_matching, Analysis, MoveDataParamEnv};
+
+use crate::MirPass;
+
+/// Removes `Drop` and `DropAndReplace` terminators whose target is known to be uninitialized at
+/// that point.
+///
+/// This is redundant with drop elaboration, but we need to do it prior to const-checking, and
+/// running const-checking after drop elaboration makes it opimization dependent, causing issues
+/// like [#90770].
```

Comment:
Is it possible to move drop elaboration earlier, so we can still do const-checking after drop elaboration?

Also, isn't this *additionally* redundant with things done inside borrowck, which also knows to ignore those `Drop`/`DropAndReplace` terminators? It doesn't actually remove them, presumably it has an analysis to figure out that they are not needed. Is that the same analysis that is also used here?

Reply:
> Is it possible to move drop elaboration earlier, so we can still do const-checking after drop elaboration?

How is that better than the status quo? To be clear, I think the status quo is perfectly valid, it just needed a stronger warning about what is allowed as part of `post_borrowck_cleanup`.

> Also, isn't this additionally redundant with things done inside borrowck, which also knows to ignore those Drop/DropAndReplace terminators? It doesn't actually remove them, presumably it has an analysis to figure out that they are not needed. Is that the same analysis that is also used here?

My long term goal ([as I stated in this comment](https://github.com/rust-lang/rust/issues/73255#issuecomment-961354314)) is to move `remove_uninit_drops` prior to borrow checking and remove the init logic in the borrow checker. I'm a little nervous about this, since I don't fully understand the purpose of the `FalseEdge`/`FalseUnwind` terminators, which the borrow checker requires . `FalseEdge`s seem to be used to connect distinct match arms to one another, which means we lose information about what the active variant is in a match arm body.

I requested some input about this idea from the MIR borrowck implementors as part of that comment. Perhaps you have some thoughts as well? It's trivial to (optionally) ignore `FalseEdge`s while computing `MaybeInitializedPlaces`, but I don't feel comfortable doing so unilaterally.

Reply:
> How is that better than the status quo?

My thinking was: if we can do drop elaboration early enough, maybe we don't have to redundantly also run remove_uninit_drops.

> I'm a little nervous about this, since I don't fully understand the purpose of the FalseEdge/FalseUnwind terminators, which the borrow checker requires . FalseEdges seem to be used to connect distinct match arms to one another, thus losing information about what the active variant is in a match arm body.

My rough idea for their purpose is that they make the borrow checker accept code that we do not want to accept -- the extra edges make variables 'more live', so that we do not accept code which depends on details such as the exact order in which disjoint match arms are being generated in the MIR. However, I do not have a good understanding of the borrow checker, so I hope you can get better info the the borrowck people -- @nikomatsakis might be able to help or at least say who should know these details. :)  @lqd and @matthewjasper  might also know more.